File: hs/src/Q400/Q456.hs
module Q400.Q456
(find132patternPlanA
    ) where

find132patternPlanA :: [Int] -> Bool
find132patternPlanA [] = False
find132patternPlanA [_] = False
find132patternPlanA [_,_] = False
find132patternPlanA nums = not . null . filter (\(x,y,z) -> x < z && z < y) . plana $ nums

plana :: [Int] -> [(Int,Int,Int)]
plana (x:y:[]) = []
plana nums = (ff nums) ++ (plana (tail nums))
                where
                    ff (a:b:[]) = []
                    ff (a:b:c) = [(i,j,k) | i <- [a], j <-[b],k <-c] ++ (ff (a:c))

File: golang/q400/q456.go
package q400

func find132patternPlanA(nums []int) bool {
	if len(nums) <= 2 {
		return false
	}
	length := len(nums)
	for i := 0; i < length; i++ {
		for j := i + 1; j < length; j++ {
			if nums[j] > (nums[i] + 1) {
				for k := j + 1; k < length; k++ {
					if nums[i] < nums[k] && nums[k] < nums[j] {
						return true
					}
				}
			}
		}
	}
	return false
}
